mmdet/models/dense_heads/ddod_head.py

```diff
@@ -0,0 +1,621 @@
+                pos_bbox_pred = bbox_pred[pos_inds]
+                pos_anchors = anchors[pos_inds]
+
+                pos_decode_bbox_pred = self.bbox_coder.decode(
+                    pos_anchors, pos_bbox_pred)
+                pos_decode_bbox_targets = self.bbox_coder.decode(
+                    pos_anchors, pos_bbox_targets)
+
+                # regression loss
+                loss_bbox = self.loss_bbox(
+                    pos_decode_bbox_pred,
+                    pos_decode_bbox_targets,
+                    avg_factor=num_total_samples)
+
+                iou_targets[pos_inds] = bbox_overlaps(
+                    pos_decode_bbox_pred.detach(),
+                    pos_decode_bbox_targets,
+                    is_aligned=True)
+                loss_iou = self.loss_iou(
+                    iou_pred,
+                    iou_targets,
+                    iou_weights,
+                    avg_factor=num_total_samples)
+            else:
+                loss_bbox = bbox_pred.sum() * 0
+                loss_iou = iou_pred.sum() * 0
+
+            return reweight_factor * loss_bbox, reweight_factor * loss_iou
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'iou_preds'))
+    def loss(self,
+             cls_scores,
+             bbox_preds,
+             iou_preds,
+             gt_bboxes,
+             gt_labels,
+             img_metas,
+             gt_bboxes_ignore=None):
+        """Compute losses of the head.
+
+        Args:
+            cls_scores (list[Tensor]): Box scores for each scale level
+                Has shape (N, num_anchors * num_classes, H, W)
+            bbox_preds (list[Tensor]): Box energies / deltas for each scale
+                level with shape (N, num_anchors * 4, H, W)
+            iou_preds (list[Tensor]): Score factor for all scale level,
+                each is a 4D-tensor, has shape (batch_size, 1, H, W).
+            gt_bboxes (list[Tensor]): Ground truth bboxes for each image with
+                shape (num_gts, 4) in [tl_x, tl_y, br_x, br_y] format.
+            gt_labels (list[Tensor]): class indices corresponding to each box
+            img_metas (list[dict]): Meta information of each image, e.g.,
+                image size, scaling factor, etc.
+            gt_bboxes_ignore (list[Tensor] | None): specify which bounding
+                boxes can be ignored when computing the loss.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        featmap_sizes = [featmap.size()[-2:] for featmap in cls_scores]
+        assert len(featmap_sizes) == self.prior_generator.num_levels
+
+        device = cls_scores[0].device
+        anchor_list, valid_flag_list = self.get_anchors(
+            featmap_sizes, img_metas, device=device)
+        label_channels = self.cls_out_channels if self.use_sigmoid_cls else 1
+
+        # classification branch assigner
+        cls_reg_targets = self.get_targets(
```

Comment:
Due to the existence of two assigner processes, there is duplicate logic inside. I don't know whether to remove it for the time being? Once removed the changes are bigger. Need another reviewer to see.